### DISCUSSION TITLE
API query to connect tasks and time sheets

### DISCUSSION THREAD

**Discussion by roskakori** [kimai/kimai#4188] [Plugins]
In the user interface of Kimai 2, each tasks has a list of the time sheets booked on it. However, I could not figure out how to connect tasks and time sheets when using the API.

My expectation would have been, that the `timesheet`API route includes a field `task_id` that has an `int` referring to a task or is `null` in case it was booked without a task.

But when I go to the "empty demo" and create a single customer, project, activity, task and then a time sheet connected to this task, the results come up as seen below.

How how are time sheets connected to tasks in the API?

Query for time sheets:
```bash
curl -X 'GET' \
  'https://demo-empty.kimai.org/api/timesheets' \
  -H 'accept: application/json' \
  -H 'X-AUTH-USER: susan_super' \
  -H 'X-AUTH-TOKEN: TODO'
```
```JSON
[
  {
    "activity": 1,
    "project": 1,
    "user": 8,
    "tags": [],
    "id": 1,
    "begin": "2023-07-14T16:36:00+0200",
    "end": "2023-07-14T16:37:00+0200",
    "duration": 60,
    "description": null,
    "rate": 1.6167,
    "internalRate": 1.6167,
    "exported": false,
    "billable": true,
    "metaFields": []
  }
]
```
Query for tasks:
```sh
curl -X 'GET' \
  'https://demo-empty.kimai.org/api/tasks' \
  -H 'accept: application/json' \
  -H 'X-AUTH-USER: susan_super' \
  -H 'X-AUTH-TOKEN: TODO'
```
```JSON
[
  {
    "tags": [],
    "activeTimesheets": [],
    "id": 1,
    "title": "Demo Task",
    "status": "progress",
    "todo": null,
    "description": null,
    "project": {
      "id": 1,
      "customer": {
        "id": 1,
        "name": "Demo Customer",
        "number": "0001",
        "comment": null,
        "visible": true,
        "billable": true,
        "color": null
      },
      "name": "Demo Project",
      "comment": null,
      "visible": true,
      "billable": true,
      "globalActivities": true,
      "color": null
    },
    "activity": {
      "id": 1,
      "project": null,
      "name": "Demo Activity",
      "comment": null,
      "visible": true,
      "billable": true,
      "color": null
    },
    "user": null,
    "team": null,
    "end": "2023-07-29T00:00:00+0200",
    "estimation": null
  }
]
```

**Comment by kevinpapst** [comment]
You can start and stop tasks, which will link the new timesheet to the task.

The task is the leading object here and only the task knows which timesheets are linked to it, not vice versa.

**Reply by roskakori** [reply]
Thanks for the clarification and the insights into the underlying data model.

However the question remains how the list of time sheets attached to the task can be queried using the API. With the new understanding I would have expected a API route like `/api/tasks/{id}/timesheets`. Sadly I can not find such a route in the API documentation (see screenshot below).

Am I correct in my understanding that at this time there is no option to query the API for the connection between tasks and time sheets?

![kimai api taks](https://github.com/kimai/kimai/assets/328726/6409c4f7-e9df-4615-b0e9-9c834088788b)

**Reply by kevinpapst** [reply]
Seems to be correct. I haven't touched that API in a while.
Please open a feature request at https://github.com/kevinpapst/kimai-plugins and I will add it to the roadmap.

**Reply by kevinpapst** [reply]
You can get the latest version including a route to "fetch the timesheets for a task" at 
https://license.kimai.org/

**Reply by roskakori** [reply]
Cool, thanks!
